### PR TITLE
[ticket/14269] Use http_exceptions in the installer instead of die()

### PR DESCRIPTION
**Pull request by CHItA**
[PHPBB3-14269](https://tracker.phpbb.com/browse/PHPBB3-14269)

### CODE REVIEW

File: phpBB/language/en/install.php [hunk 1]
```diff
@@ -82,6 +82,9 @@
 	<p>Only those databases supported on your server will be displayed.',
 
 	'ACP_LINK'	=> 'Take me to <a href="%1$s">the ACP</a>',
+
+	'INSTALL_PHPBB_IS_ALREADY_INSTALLED'	=> 'phpBB is already installed.',
```
Comment (marc1706) [inline phpBB/language/en/install.php]: How about just using INSTALL_PHPBB_ALREADY_INSTALLED and INSTALL_PHPBB_NOT_INSTALLED?
The "is" is kind of useless.

File: phpBB/phpbb/install/event/kernel_exception_subscriber.php [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+		$this->language				= $language;
+		$this->template				= $template;
+	}
+
+	/**
+	 * This listener is run when the KernelEvents::EXCEPTION event is triggered
+	 *
+	 * @param GetResponseForExceptionEvent	$event
+	 */
+	public function on_kernel_exception(GetResponseForExceptionEvent $event)
```
Comment (marc1706) [inline phpBB/phpbb/install/event/kernel_exception_subscriber.php]: Similar to the above use of "is", the "on" is IMHO useless and doesn't add any info.

File: phpBB/phpbb/install/event/kernel_exception_subscriber.php [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+		{
+			$response->setStatusCode($exception->getStatusCode());
+			$response->headers->add($exception->getHeaders());
+		}
+
+		$event->setResponse($response);
+	}
+
+	/**
+	 * Returns an array of the events the object is subscribed to
```
Comment (marc1706) [inline phpBB/phpbb/install/event/kernel_exception_subscriber.php]: an array of events the object

**Comment by marc1706** [comment]
Rest looks good :+1:

File: phpBB/language/en/install.php (hunk 1 above)
Comment (VSEphpbb) [inline phpBB/language/en/install.php]: How 'bout just INSTALL_PHPBB_INSTALLED

File: phpBB/phpbb/install/event/kernel_exception_subscriber.php (hunk 2 above)
Comment (CHItA) [inline phpBB/phpbb/install/event/kernel_exception_subscriber.php]: that's the naming convention and AFAIK not just in phpBB.

**Comment by CHItA** [comment]
@phpbb/development-team anyone has something to add about HTTP response code?

File: phpBB/phpbb/install/controller/install.php [hunk 4]
```diff
@@ -97,12 +98,14 @@ public function __construct(helper $helper, factory $factory, navigation_provide
...
 	 * @return Response|StreamedResponse
+	 *
+	 * @throws http_exception When phpBB is already installed
 	 */
 	public function handle()
 	{
 		if ($this->install_helper->is_phpbb_installed())
 		{
-			die ('phpBB is already installed');
+			throw new http_exception(404, 'INSTALL_PHPBB_INSTALLED');
```
Comment (marc1706) [inline phpBB/phpbb/install/controller/install.php]: I'd suggest a 418 but apart from that a 403 might be better.

File: phpBB/phpbb/install/controller/update.php [hunk 5]
```diff
@@ -93,12 +94,16 @@ public function __construct(helper $controller_helper, installer $installer, ins
...
+	 * @return Response|StreamedResponse
+	 *
+	 * @throws http_exception When phpBB is not installed
 	 */
 	public function handle()
 	{
 		if (!$this->install_helper->is_phpbb_installed())
 		{
-			die ('phpBB is not installed');
+			throw new http_exception(404, 'INSTALL_PHPBB_NOT_INSTALLED');
```
Comment (marc1706) [inline phpBB/phpbb/install/controller/update.php]: See above

File: phpBB/phpbb/install/controller/install.php (hunk 4 above)
Comment (Nicofuma) [inline phpBB/phpbb/install/controller/install.php]: I like 418 ;) but yeah +1 for 403

**Comment by marc1706** [comment]
Are you planning on extending the subscriber? Otherwise we might just change this to a listener and get rid of the call to getSubscribedEvents() and then rename on_kernel_exception() to onKernelException() and it will automatically be run if you define the kernel.exception event in the tag.

**Comment by CHItA** [comment]
I think this is fine as is for now

**Comment by marc1706** [comment]
Well this will certainly reduce the complexity of the file. If we're not having more than the kernel exception in this file we should really go for a listener. It gets rid of the getSubscribedEvents() method in the file completely and will autoload the onKernelException() method.

**Comment by marc1706** [comment]
@michaelcullum for the php part

**Comment by Nicofuma** [comment]
Moving the event registration to the service definition doesn't really reduce the complexity. It just moves it to the yaml file.

**Comment by Nicofuma** [comment]
I like using subscribers because this way your class is self explained. You don't need to check another file to see if and how (which event and priority?) the event listener is registered.

**Comment by marc1706** [comment]
Ok, I'm fine with that, too. Was just a suggestion.

**Comment by michaelcullum** [comment]
I'd agree with Marc as onKernelException() is self-explanatory also I think.

**Comment by Nicofuma** [comment]
It explains the name yes, but not the priority. And even if the name is clear you can forget to update the definition but to know that you have to check 2 different files.
Actually I don't see any good reason to do that in the service definition (

**Comment by CHItA** [comment]
This is how it is everywhere in the core, this has basically no disadvantages, therefore i think consistency is the best option for now. if this really bothers someone this can be modified in an other pr.

**Comment by Nicofuma** [comment]
I completely agree with the latest message of CHItA

**Pull request merged**